File: LeetCode/cpp/567_Permutation_in_String.cpp
class Solution {
public:
	bool isSame(vector<int> v1, vector<int> v2) {
		for (int i = 0; i < v1.size(); i++) {
			if (v1[i] != v2[i]) {
				return false;
			}
		}

		return true;
	}

	bool checkInclusion(string s1, string s2) {
		if (s1.length() > s2.length()) {
			return false;
		}

		vector<int> v1(26, 0), v2(26, 0);

		for (int i = 0; i < s1.length(); i++) {
			v1[s1[i] - 'a']++;
		}

		for (int i = 0; i < s1.length(); i++) {
			v2[s2[i] - 'a']++;
		}

		int left = 0, right = s1.length() - 1;

		while (right < s2.length()) {
			if (isSame(v1, v2)) {
				return true;
			}

			v2[s2[left] - 'a']--;
			left++;
			right++;

			if (right == s2.length()) {
				break;
			}

			v2[s2[right] - 'a']++;
		}

		return false;
	}
};

File: LeetCode/java/567_Permutation_in_String.java
class Solution {
    private boolean isSame(int[] list1, int[] list2) {
        for (int i = 0; i < list1.length; i++) {
            if (list1[i] != list2[i]) {
                return false;
            }
        }

        return true;
    }

    public boolean checkInclusion(String s1, String s2) {
        if (s1.length() > s2.length()) {
            return false;
        }

        int[] list1 = new int[26];
        int[] list2 = new int[26];

        for (int i = 0; i < s1.length(); i++) {
            list1[s1.charAt(i) - 'a']++;
        }

        for (int i = 0; i < s1.length(); i++) {
            list2[s2.charAt(i) - 'a']++;
        }

        int left = 0, right = s1.length() - 1;

        while (right < s2.length()) {
            if (isSame(list1, list2)) {
                return true;
            }

            list2[s2.charAt(left) - 'a']--;
            left++;
            right++;

            if (right == s2.length()) {
                break;
            }

            list2[s2.charAt(right) - 'a']++;
        }

        return false;
    }
}